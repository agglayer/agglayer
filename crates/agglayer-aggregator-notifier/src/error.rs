/// Error type for the notifier
#[derive(Debug, thiserror::Error)]
pub enum NotifierError {
    #[error("unable to build notifier")]
    UnableToBuildNotifier,
}
