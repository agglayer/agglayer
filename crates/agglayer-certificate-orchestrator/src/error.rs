use agglayer_types::{CertificateId, Height, NetworkId, ProofVerificationError};
use pessimistic_proof::ProofError;

#[derive(thiserror::Error, Debug)]
pub enum PreCertificationError {
    #[error("Storage error: {0}")]
    Storage(#[from] agglayer_storage::error::Error),

    #[error("certificate not found for network {0} at height {1}")]
    CertificateNotFound(NetworkId, Height),
    #[error("proof already exists for network {0} at height {1} for certificate {2}")]
    ProofAlreadyExists(NetworkId, Height, CertificateId),
}
#[derive(thiserror::Error, Debug)]
pub enum CertificationError {
    #[error(
        "Failed to retrieve the trusted sequencer address for network {0} during proving phase"
    )]
    TrustedSequencerNotFound(NetworkId),
    #[error("Failed to retrieve the l1 info root for the l1 leaf count: {1} for certificate {0}")]
    L1InfoRootNotFound(CertificateId, u32),
    #[error("proof verification failed")]
    ProofVerificationFailed { source: ProofVerificationError },
    #[error("prover execution failed: {source}")]
    ProverExecutionFailed { source: ProofError },
    #[error("native execution failed: {source:?}")]
    NativeExecutionFailed { source: ProofError },
    #[error("Type error: {source}")]
    Types { source: agglayer_types::Error },
    #[error("Serialize error")]
    Serialize { source: bincode::Error },
    #[error("Deserialize error")]
    Deserialize { source: bincode::Error },
    #[error("internal error: {0}")]
    InternalError(String),
    #[error("Storage error: {0}")]
    Storage(#[from] agglayer_storage::error::Error),
}

#[derive(thiserror::Error, Debug)]
pub enum Error {
    #[error("Clock error: {0}")]
    Clock(#[from] agglayer_clock::Error),

    #[error(transparent)]
    PreCertification(#[from] PreCertificationError),
    #[error(transparent)]
    Certification(#[from] CertificationError),

    #[error("Storage error: {0}")]
    Storage(#[from] agglayer_storage::error::Error),
    #[error("internal error")]
    InternalError,

    #[error("The status of the certificate is invalid")]
    InvalidCertificateStatus,

    #[error("Failed to settle the certificate {certificate_id}: {error}")]
    SettlementError {
        certificate_id: CertificateId,
        error: String,
    },
}
