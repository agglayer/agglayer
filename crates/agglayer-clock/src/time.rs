use std::{
    num::NonZeroU64,
    sync::{
        atomic::{AtomicU64, Ordering},
        Arc,
    },
    time::Duration,
};

use chrono::{DateTime, Utc};
use tokio::{
    sync::broadcast,
    time::{interval_at, Instant},
};
use tokio_util::sync::CancellationToken;
use tracing::{debug, error};

use crate::{Clock, ClockRef, Error, Event, BROADCAST_CHANNEL_SIZE};

/// Time based [`Clock`] implementation.
///
/// Simulate blockchain block production by increasing Block height by 1 every
/// second. Epoch duration can be configured when creating the Clock.
pub struct TimeClock {
    genesis: DateTime<Utc>,
    current_block: Arc<AtomicU64>,
    epoch_duration: NonZeroU64,
    current_epoch: Arc<AtomicU64>,
}

#[async_trait::async_trait]
impl Clock for TimeClock {
    async fn spawn(mut self, cancellation_token: CancellationToken) -> Result<ClockRef, Error> {
        let (sender, _receiver) = broadcast::channel(BROADCAST_CHANNEL_SIZE);

        let clock_ref = ClockRef {
            sender: sender.clone(),
            current_epoch: self.current_epoch.clone(),
            block_height: self.current_block.clone(),
        };

        // Spawn the Clock task directly
        tokio::spawn(async move {
            self.run(sender, cancellation_token).await;
        });

        Ok(clock_ref)
    }
}

impl TimeClock {
    /// Create a new [`TimeClock`] instance based on the current datetime and an
    /// Epoch.
    pub fn new_now(epoch_duration: NonZeroU64) -> Self {
        Self::new(Utc::now(), epoch_duration)
    }

    /// Create a new [`TimeClock`] instance based on a genesis datetime and an
    /// Epoch duration.
    pub fn new(genesis: DateTime<Utc>, epoch_duration: NonZeroU64) -> Self {
        Self {
            genesis,
            current_block: Arc::new(AtomicU64::new(0)),
            epoch_duration,
            current_epoch: Arc::new(AtomicU64::new(0)),
        }
    }

    /// Run the Clock task.
    async fn run(
        &mut self,
        sender: broadcast::Sender<Event>,
        cancellation_token: CancellationToken,
    ) {
        let mut interval = interval_at(Instant::now(), Duration::from_secs(1));

        // Compute the current Block height and Epoch number
        let current_block = self.update_block_height();
        let current_epoch = self.calculate_epoch_number(current_block);

        // Use compare_exchange to ensure the initial current_epoch is 0
        if let Err(stored_value) = self.current_epoch.compare_exchange(
            0,
            current_epoch,
            Ordering::AcqRel,
            Ordering::Relaxed,
        ) {
            let error_message = format!(
                "The current_epoch has already been modified. Shutting down the Clock task. \
                 Stored value: {}",
                stored_value
            );
            #[cfg(not(test))]
            {
                error!("{}", error_message);
                std::process::exit(1);
            }
            #[cfg(test)]
            panic!("{}", error_message);
        }

        loop {
            tokio::select! {
                _ = cancellation_token.cancelled() => {
                    debug!("Clock task cancelled");
                    break;
                }
                _ = interval.tick() => {

                    // Increase the Block height by 1.
                    // The `fetch_add` method returns the previous value, so we need to add 1 to it
                    // to get the current Block height.
                    if let Some(current_block) = self
                        .current_block
                            .fetch_add(1, Ordering::Release)
                            .checked_add(1)
                    {
                        // If the current Block height is a multiple of the Epoch duration,
                        // the current Epoch has ended. In this case, we need to update the
                        // new Epoch number and send an `EpochEnded` event to the subscribers.
                        if current_block % self.epoch_duration == 0 {
                            match self.update_epoch_number() {
                                Ok(epoch_ended) => {
                                    _ = sender.send(Event::EpochEnded(epoch_ended));
                                }
                                Err((current_epoch, expected)) => {
                                    error!(
                                        "Unexpected error computing the current Epoch: current_epoch={}, \
                                        expected_epoch={}, current_block={}",
                                        current_epoch, expected, current_block
                                    );
                                    cancellation_token.cancel();
                                }
                            }
                        }
                    } else {
                       error!("Block height overflowed the u64 limit. \
                           This is an unexpected situation and could lead to unexpected behavior. \
                           Please report this issue to the developers. https://github.com/agglayer/agglayer/issues/new \
                           The node will now kill itself to prevent further damage.");

                        cancellation_token.cancel();
                    }
                }
            }
        }
    }

    /// Updates the Block height of this [`TimeClock`].
    ///
    /// This method is used to update the Block height based on the
    /// genesis datetime and the current datetime.
    ///
    /// The Block height is the number of seconds since the genesis datetime.
    fn update_block_height(&mut self) -> u64 {
        let block_height = self.calculate_block_height();

        self.current_block.store(block_height, Ordering::Release);

        block_height
    }

    /// Updates the current Epoch of this [`TimeClock`].
    ///
    /// This method is used to update the current Epoch number based on the
    /// Block height and the Epoch duration.
    ///
    /// To define the current Epoch number, the Epoch duration divides the Block
    /// height.
    fn update_epoch_number(&mut self) -> Result<u64, (u64, u64)> {
        let current_block = self.current_block.load(Ordering::Acquire);

        let current_epoch = self.calculate_epoch_number(current_block);
        let expected_epoch = current_epoch.saturating_sub(1);

        match self.current_epoch.compare_exchange(
            expected_epoch,
            current_epoch,
            Ordering::Acquire,
            Ordering::Relaxed,
        ) {
            Ok(previous) => Ok(previous),
            Err(stored) => Err((stored, expected_epoch)),
        }
    }

    /// Calculate an Epoch number based on a Block number.
    fn calculate_epoch_number(&self, from_block: u64) -> u64 {
        from_block / self.epoch_duration
    }

    /// Calculate the Block height.
    fn calculate_block_height(&self) -> u64 {
        std::cmp::max(
            Utc::now()
                .naive_utc()
                .signed_duration_since(self.genesis.naive_utc())
                .num_seconds(),
            0,
        ) as u64
    }
}

#[cfg(test)]
mod tests {
    use std::{num::NonZeroU64, sync::atomic::Ordering};

    use chrono::{Duration, Utc};
    use tokio::sync::broadcast;
    use tokio_util::sync::CancellationToken;

    use crate::{Clock, ClockRef, Event, TimeClock, BROADCAST_CHANNEL_SIZE};

    #[tokio::test]
    async fn test_time_clock() {
        let genesis = Utc::now()
            .checked_sub_signed(Duration::seconds(30))
            .unwrap();

        let clock = TimeClock::new(genesis, NonZeroU64::new(5).unwrap());

        let token = CancellationToken::new();
        let clock_ref = clock.spawn(token.clone()).await.unwrap();

        let mut recv = clock_ref.subscribe().unwrap();

        assert_eq!(recv.recv().await, Ok(Event::EpochEnded(6)));
        assert_eq!(clock_ref.current_epoch(), 7);
        assert!(clock_ref.current_block_height() >= 30);
    }

    #[tokio::test]
    async fn test_time_clock_catchup() {
        let genesis = Utc::now()
            .checked_sub_signed(Duration::seconds(30))
            .unwrap();

        let clock = TimeClock::new(genesis, NonZeroU64::new(2).unwrap());

        let token = CancellationToken::new();
        let clock_ref = clock.spawn(token.clone()).await.unwrap();

        let mut recv = clock_ref.subscribe().unwrap();
        assert_eq!(recv.recv().await, Ok(Event::EpochEnded(15)));
        assert!(recv.try_recv().is_err());
        assert_eq!(clock_ref.current_epoch(), 16);
        assert!(clock_ref.current_block_height() >= 30);
        tokio::time::sleep(std::time::Duration::from_secs(5)).await;

        assert_eq!(recv.recv().await, Ok(Event::EpochEnded(16)));
        assert_eq!(recv.recv().await, Ok(Event::EpochEnded(17)));

        assert_eq!(clock_ref.current_epoch(), 18);
        assert!(clock_ref.current_block_height() >= 35);
    }

    #[tokio::test]
    async fn test_time_clock_overflow() {
        let genesis = Utc::now()
            .checked_sub_signed(Duration::seconds(30))
            .unwrap();

        let mut clock = TimeClock::new(genesis, NonZeroU64::new(2).unwrap());
        let blocks = clock.current_block.clone();
        let (sender, _receiver) = broadcast::channel(BROADCAST_CHANNEL_SIZE);

        let clock_ref = ClockRef {
            sender: sender.clone(),
            current_epoch: clock.current_epoch.clone(),
            block_height: clock.current_block.clone(),
        };

        let token = CancellationToken::new();
        let mut fut = Box::pin(clock.run(sender, token.clone()));
        let mut recv = clock_ref.subscribe().unwrap();

        _ = futures::poll!(&mut fut);
        tokio::time::sleep(std::time::Duration::from_secs(2)).await;

        _ = futures::poll!(&mut fut);
        assert_eq!(recv.try_recv(), Ok(Event::EpochEnded(15)));

        // Overflow the block height on next poll
        blocks.store(u64::MAX - 1, std::sync::atomic::Ordering::SeqCst);

        tokio::time::sleep(std::time::Duration::from_secs(2)).await;
        _ = futures::poll!(&mut fut);

        assert!(token.is_cancelled());
    }

    #[tokio::test]
    #[should_panic]
    async fn test_initial_epoch_update_error() {
        let genesis = Utc::now()
            .checked_sub_signed(Duration::seconds(30))
            .unwrap();

        let mut clock = TimeClock::new(genesis, NonZeroU64::new(5).unwrap());
        clock.current_epoch.store(1, Ordering::Relaxed);

        let (sender, _receiver) = broadcast::channel(BROADCAST_CHANNEL_SIZE);

        let token = CancellationToken::new();
        let _ = clock.run(sender, token).await;
    }
}
