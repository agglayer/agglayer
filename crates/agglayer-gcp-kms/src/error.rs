//! The [`enum@Error`] enum represents errors that can occur in the KMS
//! operations.
//!
//! It includes errors from the KMS provider and configuration errors.

use ethers_gcp_kms_signer::CKMSError;
use thiserror::Error;

/// Represents errors that can occur in the KMS operations.
#[derive(Debug, Error)]
pub enum Error {
    /// An error occurred with the KMS provider.
    ///
    /// This variant wraps the underlying [`CKMSError`] from the
    /// `ethers_gcp_kms_signer` library.
    #[error("KMS Provider error: {0}")]
    KmsProvider(#[from] CKMSError),

    /// An error occurred with the KMS configuration.
    ///
    /// This variant is used when a required key or environment variable is
    /// missing.
    #[error("KMS configuration error: missing key or env {0}")]
    KmsConfig(&'static str),
}
