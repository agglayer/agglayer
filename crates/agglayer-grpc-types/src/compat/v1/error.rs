use std::fmt;

use agglayer_types::primitives::SignatureError;
use tonic_types::FieldViolation;

#[derive(Debug, thiserror::Error)]
pub enum SourceError {
    #[error(transparent)]
    Bincode(#[from] bincode::Error),

    #[error(transparent)]
    Signature(#[from] SignatureError),
}

#[derive(Clone, Copy, Debug)]
pub enum ErrorKind {
    InvalidData,
    MissingField,
}

impl fmt::Display for ErrorKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ErrorKind::InvalidData => {
                write!(f, "Invalid data")
            }
            ErrorKind::MissingField => {
                write!(f, "Missing field")
            }
        }
    }
}

#[derive(Debug, thiserror::Error)]
pub struct Error {
    kind: ErrorKind,
    message: String,
    field: Vec<&'static str>,
    #[source]
    source: Option<SourceError>,
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        if !self.field.is_empty() {
            write!(f, "{}: ", self.field_str())?;
        }
        write!(f, "{}", self.message)
    }
}

impl Error {
    pub fn missing_field(f: &'static str) -> Self {
        Error {
            kind: ErrorKind::MissingField,
            message: "required field is missing".to_string(),
            field: vec![f],
            source: None,
        }
    }

    pub fn invalid_data(m: String) -> Self {
        Error {
            kind: ErrorKind::InvalidData,
            message: m,
            field: vec![],
            source: None,
        }
    }

    pub fn inside_field(mut self, f: &'static str) -> Self {
        self.field.push(f);
        self.field.rotate_right(1);
        self
    }

    pub fn serializing_proof(e: bincode::Error) -> Self {
        Error {
            kind: ErrorKind::InvalidData,
            message: "failed to serialize proof".to_string(),
            field: vec![],
            source: Some(SourceError::Bincode(e)),
        }
    }

    pub fn deserializing_proof(e: bincode::Error) -> Self {
        Error {
            kind: ErrorKind::InvalidData,
            message: "failed to deserialize proof".to_string(),
            field: vec![],
            source: Some(SourceError::Bincode(e)),
        }
    }

    pub fn parsing_signature(e: SignatureError) -> Self {
        Error {
            kind: ErrorKind::InvalidData,
            message: "failed to parse signature".to_string(),
            field: vec![],
            source: Some(SourceError::Signature(e)),
        }
    }

    pub fn kind(&self) -> ErrorKind {
        self.kind
    }

    pub fn field(&self) -> &[&'static str] {
        &self.field
    }

    pub fn field_str(&self) -> String {
        if self.field.is_empty() {
            ".".to_string()
        } else {
            self.field.join(".")
        }
    }
}

impl From<&Error> for Vec<FieldViolation> {
    fn from(value: &Error) -> Self {
        vec![FieldViolation::new(
            value.field_str(),
            value.message.clone(),
        )]
    }
}
