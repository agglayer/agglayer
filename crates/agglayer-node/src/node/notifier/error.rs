#[derive(Debug, thiserror::Error)]
pub(crate) enum NotifierError {
    #[error("unable to build notifier")]
    UnableToBuildNotifier,
}
