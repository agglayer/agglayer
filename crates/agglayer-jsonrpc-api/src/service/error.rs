//! Error types for the top-level Agglayer service.

use agglayer_contracts::contracts::{
    PolygonRollupManager::PolygonRollupManagerErrors, PolygonZkEvm::PolygonZkEvmErrors,
};
use agglayer_rate_limiting::RateLimited as RateLimitedError;
use agglayer_rpc::error::SignatureVerificationError;
pub use agglayer_types::Digest;
use alloy::{contract::Error as ContractError, primitives::B256};

pub use crate::kernel::{CheckTxStatusError, SettlementError, ZkevmNodeVerificationError};

#[derive(Debug, thiserror::Error)]
pub enum CertificateRetrievalError {
    #[error("Data for certificate {certificate_id} not found")]
    NotFound { certificate_id: Digest },
}

#[derive(Debug, thiserror::Error)]
pub enum TxStatusError {
    #[error("Status retrieval error: {0}")]
    StatusCheck(CheckTxStatusError),

    #[error("Failed to get L1 block: {0}")]
    L1BlockRetrieval(CheckTxStatusError),

    #[error("Transaction {hash} not found")]
    TxNotFound { hash: B256 },
}

#[derive(Debug, thiserror::Error)]
pub enum SendTxError {
    #[error("Rate limited: {0}")]
    RateLimited(#[from] RateLimitedError),

    #[error(transparent)]
    SignatureError(#[from] SignatureVerificationError),

    #[error("Rollup {rollup_id} not registered")]
    RollupNotRegistered { rollup_id: u32 },

    #[error("Zkevm error during dry run")]
    DryRunZkEvm(PolygonZkEvmErrors),

    #[error("Rollup manager error during dry run")]
    DryRunRollupManager(PolygonRollupManagerErrors),

    #[error("Error during dry run: {0}")]
    DryRunOther(ContractError),

    #[error("Failed to verify local exit root or state root: {0}")]
    RootVerification(ZkevmNodeVerificationError),

    #[error("Settlement failed: {0}")]
    Settlement(SettlementError),
}

impl SendTxError {
    /// Decode the dry run contract errors.
    pub fn dry_run(err: ContractError) -> Self {
        // Note: In alloy, contract error decoding is handled differently
        // This is a simplified version and may need adjustment based on actual contract
        // error handling
        Self::DryRunOther(err)
    }
}

impl From<SettlementError> for SendTxError {
    fn from(err: SettlementError) -> Self {
        match err {
            SettlementError::RateLimited(e) => e.into(),
            e => Self::Settlement(e),
        }
    }
}
