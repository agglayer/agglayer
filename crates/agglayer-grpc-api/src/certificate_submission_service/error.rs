use std::collections::HashMap;

use agglayer_grpc_types::node::v1::SubmitCertificateErrorKind;
use agglayer_rpc::CertificateSubmissionError;
use tonic_types::{ErrorDetails, StatusExt as _};
use tracing::{error, warn};

pub(crate) struct CertificateSubmissionErrorWrapper {
    inner: CertificateSubmissionError,
    context: &'static str,
}

impl CertificateSubmissionErrorWrapper {
    pub(crate) fn new(inner: CertificateSubmissionError, context: &'static str) -> Self {
        Self { inner, context }
    }
}

impl From<CertificateSubmissionErrorWrapper> for tonic::Status {
    fn from(error: CertificateSubmissionErrorWrapper) -> Self {
        match error.inner {
            agglayer_rpc::CertificateSubmissionError::Storage(error) => {
                error!(?error, "returning internal storage error to RPC");
                tonic::Status::internal("Internal storage error")
            }

            agglayer_rpc::CertificateSubmissionError::OrchestratorNotResponsive => {
                warn!("returning orchestrator not responsive to RPC");
                tonic::Status::internal("Orchestrator not responsive")
            }

            agglayer_rpc::CertificateSubmissionError::SignatureError(
                signature_verification_error,
            ) => tonic::Status::with_error_details(
                tonic::Code::InvalidArgument,
                "Signature verification error",
                ErrorDetails::with_error_info(
                    SubmitCertificateErrorKind::SignatureVerification.as_str_name(),
                    error.context,
                    [("error".into(), format!("{signature_verification_error:?}"))],
                ),
            ),

            agglayer_rpc::CertificateSubmissionError::UnableToReplacePendingCertificate {
                reason,
                height,
                network_id,
                stored_certificate_id,
                replacement_certificate_id,
                source,
            } => {
                let mut details: HashMap<String, String> = vec![
                    ("reason".into(), reason),
                    ("height".into(), height.to_string()),
                    ("network_id".into(), network_id.to_string()),
                    (
                        "stored_certificate_id".into(),
                        stored_certificate_id.to_string(),
                    ),
                    (
                        "replacement_certificate_id".into(),
                        replacement_certificate_id.to_string(),
                    ),
                ]
                .into_iter()
                .collect();

                if let Some(source) = source {
                    details.insert("error".into(), format!("{source:?}"));
                }

                tonic::Status::with_error_details(
                    tonic::Code::InvalidArgument,
                    "Unable to replace pending certificate",
                    ErrorDetails::with_error_info(
                        SubmitCertificateErrorKind::UnableToReplacePendingCertificate.as_str_name(),
                        error.context,
                        details,
                    ),
                )
            }
        }
    }
}
