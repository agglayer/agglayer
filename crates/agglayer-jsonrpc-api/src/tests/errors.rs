//! Tests for rendering RPC errors

use std::time::Duration;

use agglayer_rate_limiting::{self, component, Component};
use agglayer_rpc::error::SignatureVerificationError;
use agglayer_types::{Address, CertificateId, Digest};
use alloy::{
    contract::Error as ContractError,
    primitives::{SignatureError as AlloySignatureError, B256},
    signers::k256,
};
use jsonrpsee::types::ErrorObjectOwned;

use crate::{
    kernel::{self, ZkevmNodeVerificationError},
    service, Error,
};

// Update type aliases to use alloy types
type CheckTxStatusError = kernel::CheckTxStatusError;
type SendTxError = service::SendTxError;
type SettlementError = kernel::SettlementError;
type TxStatusError = service::TxStatusError;
type WallClockLimitedInfo = <component::SendTx as Component>::LimitedInfo;

#[rstest::rstest]
#[case("rollup_not_reg", SendTxError::RollupNotRegistered { rollup_id: 1337 })]
#[case(
    "sig_invalid_len",
    SendTxError::SignatureError(SignatureVerificationError::CouldNotRecoverTxSigner(
        AlloySignatureError::FromHex(alloy::hex::FromHexError::InvalidStringLength)
    ))
)]
#[case("sig_verif", SendTxError::SignatureError(SignatureVerificationError::InvalidSigner {
    signer: Address::from([0x11; 20]),
    trusted_sequencer: Address::from([0x22; 20]),
}))]
#[case(
    "sig_recov",
    SendTxError::SignatureError(SignatureVerificationError::CouldNotRecoverTxSigner(
        AlloySignatureError::K256(k256::ecdsa::Error::new())
    ))
)]
#[case(
    "cert_sig",
    SendTxError::SignatureError(SignatureVerificationError::CouldNotRecoverCertSigner(
        agglayer_types::SignerError::Recovery(AlloySignatureError::K256(
            k256::ecdsa::Error::new()
        ))
    ))
)]
#[case(
    "sig_signer",
    SendTxError::SignatureError(SignatureVerificationError::InvalidSigner{
        signer: Address::from([0x33; 20]),
        trusted_sequencer: Address::from([0x44; 20]),
    })
)]
#[case(
    "sig_contract",
    SendTxError::SignatureError(SignatureVerificationError::ContractError(
        alloy::contract::Error::ContractNotDeployed
    ))
)]
#[case("dry_run_rollup_man", SendTxError::DryRunRollupManager(
    agglayer_contracts::contracts::PolygonRollupManager::PolygonRollupManagerErrors::FinalNumBatchBelowLastVerifiedBatch(
        agglayer_contracts::contracts::PolygonRollupManager::FinalNumBatchBelowLastVerifiedBatch {}
    )
))]
#[case(
    "root_bad_rollup",
    SendTxError::RootVerification(ZkevmNodeVerificationError::InvalidRollupId(13))
)]
#[case(
    "root_rpc",
    SendTxError::RootVerification(ZkevmNodeVerificationError::RpcError(
        jsonrpsee::core::client::Error::Custom("Node smells too much".into())
    ))
)]
#[case(
    "root_state",
    SendTxError::RootVerification(ZkevmNodeVerificationError::InvalidStateRoot {
        expected: B256::from([0x55; 32]),
        got: B256::from([0x66; 32]),
    })
)]
#[case(
    "root_exit",
    SendTxError::RootVerification(ZkevmNodeVerificationError::InvalidExitRoot {
        expected: B256::from([0x77; 32]),
        got: B256::from([0x88; 32]),
    })
)]
#[case("settle_receipt", SendTxError::Settlement(SettlementError::NoReceipt))]
#[case(
    "settle_io",
    SendTxError::Settlement(SettlementError::ProviderError(alloy::transports::RpcError::Transport(
        alloy::transports::TransportErrorKind::Custom("Settlement transport error".to_string().into())
    ))
))]
#[case(
    "settle_contract",
    SendTxError::Settlement(SettlementError::ContractError(ContractError::TransportError(
        alloy::transports::RpcError::Transport(
            alloy::transports::TransportErrorKind::Custom("Contract transport error".to_string().into())
        )
    )))
)]
#[case(
    "settle_l1_timeout",
    SendTxError::Settlement(SettlementError::Timeout(Duration::from_secs(30 * 60)))
)]
#[case(
    "rate_disallowed",
    SendTxError::RateLimited(agglayer_rate_limiting::RateLimited::SendTxDiabled {})
)]
#[case(
    "rate_sendtx",
    SendTxError::RateLimited(agglayer_rate_limiting::RateLimited::SendTxRateLimited(WallClockLimitedInfo {
        max_per_interval: 3,
        time_interval: Duration::from_secs(30 * 60),
        until_next: Some(Duration::from_secs(123)),
    }))
)]
#[case(
    "rate_sendtx_nonext",
    SendTxError::RateLimited(agglayer_rate_limiting::RateLimited::SendTxRateLimited(WallClockLimitedInfo {
        max_per_interval: 4,
        time_interval: Duration::from_secs(40 * 60),
        until_next: None,
    }))
)]
#[case(
    "txstatus_notfound",
    TxStatusError::TxNotFound { hash: B256::from([0x97; 32]) }
)]
#[case(
    "txstatus_check",
    TxStatusError::StatusCheck(CheckTxStatusError::ProviderError(
        alloy::transports::RpcError::Transport(
            alloy::transports::TransportErrorKind::Custom("Signer unavailable".to_string().into())
        )
    ))
)]
#[case(
    "cert_notfound",
    agglayer_rpc::CertificateRetrievalError::NotFound { certificate_id: CertificateId::new(Digest([0x51; 32])) }
)]
fn rpc_error_rendering(#[case] name: &str, #[case] err: impl Into<Error>) {
    let err: Error = err.into();
    let debug_str = format!("{err:?}");
    let err_obj = ErrorObjectOwned::from(err);
    let err_json_string = {
        // Going through an extra encode/decode here helps normalize the output.
        let json_string = serde_json::to_string(&err_obj).unwrap();
        let json = serde_json::from_str::<serde_json::Value>(&json_string).unwrap();
        serde_json::to_string_pretty(&json).unwrap()
    };

    insta::assert_snapshot!(name, err_json_string, &debug_str);
}
