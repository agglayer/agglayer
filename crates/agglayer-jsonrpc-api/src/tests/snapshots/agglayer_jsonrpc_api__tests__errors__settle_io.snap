---
source: crates/agglayer-jsonrpc-api/src/tests/errors.rs
expression: "Settlement(IoError(\"Settlement transport error\"))"
---
{
  "code": -10004,
  "data": {
    "settlement": {
      "io-error": "Settlement transport error"
    }
  },
  "message": "L1 settlement error: IO error"
}
